src/python/pants/engine/internals/graph.py

```diff
@@ -135,37 +134,74 @@ async def resolve_targets_with_origins(
+
+    def visit(address: Address):
+        if address in visited:
+            # NB: File-level dependencies are cycle tolerant.
+            if not address.generated_base_target_name and address in path_stack:
+                raise CycleException(address, tuple(path_stack) + (address,))
+            return
+        path_stack.add(address)
+        visited.add(address)
+
+        for dep_address in dependencies[address]:
+            visit(dep_address)
+
+        path_stack.remove(address)
+
+    for root in roots:
+        visit(root)
+        if path_stack:
+            raise AssertionError(
+                f"The stack of visited nodes should have been empty at the end of recursion, "
+                f"but it still contained: {path_stack}"
+            )
 
 
 @rule
-async def transitive_targets(addresses: Addresses) -> TransitiveTargets:
-    """Given Addresses, kicks off recursion on expansion of TransitiveTargets.
+async def transitive_targets(targets: Targets) -> TransitiveTargets:
+    """Find all the targets transitively depended upon by the target roots.
 
-    The TransitiveTarget dataclass represents a structure-shared graph, which we walk and flatten
-    here. The engine memoizes the computation of TransitiveTarget, so when multiple
-    TransitiveTargets objects are being constructed for multiple roots, their structure will be
-    shared.
+    This uses iteration, rather than recursion, so that we can tolerate dependency cycles. Unlike a
+    traditional BFS algorithm, we batch each round of traversals via `MultiGet` for improved
+    performance / concurrency.
     """
-    tts = await MultiGet(Get(TransitiveTarget, Address, a) for a in addresses)
+    visited: OrderedSet[Target] = OrderedSet()
+    queued = FrozenOrderedSet(targets)
+    dependency_mapping: Dict[Address, Tuple[Address, ...]] = {}
+    while queued:
+        direct_dependencies = await MultiGet(
+            Get(Targets, DependenciesRequest(tgt.get(Dependencies))) for tgt in queued
+        )
 
-    dependencies: OrderedSet[Target] = OrderedSet()
-    to_visit = deque(itertools.chain.from_iterable(tt.dependencies for tt in tts))
-    while to_visit:
-        tt = to_visit.popleft()
-        if tt.root in dependencies:
-            continue
-        dependencies.add(tt.root)
-        to_visit.extend(tt.dependencies)
+        dependency_mapping.update(
+            zip(
+                (t.address for t in queued),
+                (tuple(t.address for t in deps) for deps in direct_dependencies),
+            )
+        )
+
+        queued = FrozenOrderedSet(itertools.chain.from_iterable(direct_dependencies)).difference(
+            visited
+        )
+        visited.update(queued)
 
-    return TransitiveTargets(tuple(tt.root for tt in tts), FrozenOrderedSet(dependencies))
+    transitive_targets = TransitiveTargets(tuple(targets), FrozenOrderedSet(visited))
+    _detect_cycles(tuple(t.address for t in targets), dependency_mapping)
```

Comment:
I suspect running DFS has a real performance impact on mission-critical code. But I also don't know of another way to detect these cycles at the same time as we're resolvings deps via BFS..

Any idea how bad the impact is when running `./pants dependencies --transitive ::`?

One thing Benjy and I started to discuss this morning is whether to have a global flag `--tolerate-cycles`. Gordon gave feedback that he wouldn't want to tolerate cycles with dep inference / generated subtargets either. We would only run this code if you have `--no-tolerate-cycles`.

Reply:
It appears to add about 4.8% to the runtime.

> One thing Benjy and I started to discuss this morning is whether to have a global flag --tolerate-cycles.

I'd prefer something that could be toggled on a target by target basis.

Reply:
I think that would be fine with me to be a target by target basis. I'd rather have that than always no matter what tolerating generated subtargets.

Reply:
As long as there's a way to opt-in to (or out of) enforcing cycle hygiene.